src/main/scala/is/hail/annotations/MemoryBlock.scala

```diff
@@ -187,19 +248,367 @@ final class MemoryBuffer(sizeHint: Int = 128) {
     offset = (offset + (alignment - 1)) & ~(alignment - 1)
   }
 
-  def copyFrom(other: MemoryBlock, readStart: Int, writeStart: Int, size: Int) {
+  def copyFrom(other: MemoryBuffer, readStart: Int, writeStart: Int, size: Int) {
     assert(writeStart <= (offset - size))
-    mb.copyFrom(other, readStart, writeStart, size)
+    mb.copyFrom(other.mb, readStart, writeStart, size)
   }
 
   def clear() {
     offset = 0
   }
 
+  def copy(): MemoryBuffer = new MemoryBuffer(mb.copy((offset + 7) / 8))
+
   def result(): MemoryBlock = {
     val reqLength = (offset + 7) / 8
     val arr = new Array[Long](reqLength)
     Platform.copyMemory(mb.mem, Platform.LONG_ARRAY_OFFSET, arr, Platform.LONG_ARRAY_OFFSET, offset)
     new MemoryBlock(arr)
   }
 }
+
+case class RegionValue(region: MemoryBuffer, offset: Int)
+
+class RegionValueBuilder(region: MemoryBuffer) {
+  var start: Int = _
+
+  val typestk = new mutable.Stack[Type]()
```

Comment:
I know.  Odersky should be ashamed.  Profiling didn't turn this up to be a hot spot (yet), so I left it for now.